README
NAME
    Catalyst::Model::SVN - Catalyst Model to browse Subversion repositories

SYNOPSIS
        # Model
        __PACKAGE__->config(
            repository => '/path/to/svn/root/or/path'
        );

        # Controller
        sub default : Private {
            my ($self, $c) = @_;
            my $path = join('/', $c->req->args);
            my $revision = $c->req->param('revision') || 'HEAD';

            $c->stash->{'repository_revision'} = MyApp::M::SVN->revision;
            $c->stash->{'items'} = MyApp::M::SVN->ls($path, $revision);

            $c->stash->{'template'} = 'blog.tt';
        };

DESCRIPTION
    This model class uses the perl-subversion bindings to access a
    Subversion repository and list items and view their contents. It is
    currently only a read-only client but may expand to be a fill fledged
    client at a later time.

CONFIG
    The following configuration options are available:

  repository
    Returns a URI object of the full path to the root of, or any directory
    in your Subversion repository. This can be one of http://, svn://, or
    file:/// schemes.

    This value comes from the config key 'repository'.

  revision
    This is the default revision to use when no revision is specified. By
    default, this will be "HEAD".

METHODS
  cat($path [, $revision])
    Returns the contents of the path specified. If "path" is a copy, the
    logs are transversed to find original. The request is then reissued for
    the original path for the "revision" specified.

  ls($path [, $revision])
    Returns a array of Catalyst::Model::SVN::Item objects in list context,
    each representing an entry in the specified repository path. In scalar
    context, it returns an array reference. If "path" is a copy, the logs
    are transversed to find the original. The request is then reissued for
    the original path for the "revision" specified.

  propget($path, $propname [, $revision])
    Returns a specific property for a path at a specified revision name.

    Note: This method is inefficient, if you want to extract multiple
    properties of a single item then use the props_hr method.

  properties_hr($path [, $revision])
    Returns a reference to a hash with all the properties set on an object
    at a specific revision.

  repository
    Returns the repository specified in the configuration "repository"
    option.

  revision
    Returns the latest revisions of the repository you are connected to.

SEE ALSO
    Catalyst::Manual, Catalyst::Helper, Catalyst::Model::SVN::Item, SVN::Ra

AUTHORS
        Christopher H. Laco
        CPAN ID: CLACO
        claco@chrislaco.com
        http://today.icantfocus.com/blog/
    
        Tomas Doran
        CPAN ID: BOBTFISH
        bobtfish@bobtfish.net
    


Changes
Revision history for Perl module Catalyst::Model::SVN

0.12 Mon May 26 09:55:13 2008
   - Build module with ShipIt
   - Fix bug causing freeze when no config was supplied + test.
   - Add extra debugging when $ENV{DEBUG_TEST} is set.

0.11 Sat Dec 29 21:27:51 2007
   - Fix getting the config from outside the class (i.e. config from an    
     external config file now works). No tests yet, but seems to work.

0.10 Sat Dec 8 01:18:05 2007
   - Fix a nasty bug causing SVN::Ra to throw an exception in some versions (i.e. on my Debian box)

0.09 Fri Dec 7 11:48:27 2007
    - Fix crash when calling ls for a directory with a trailing / in the path 
      requested which occurs when your SVn server is over HTTP and below the 
      root directory.

0.08 Thu Dec 6 18:35:57 2007
    - Make _repos_path collapse URLs with multiple // in them. Tests for this.
    - Stop warnings being fatal in the Makefile, this breaks me on some perls.
    - Add propget as a method on Model::SVN, as well as on ::Items. Tests for this.
    - Test / fix content regex I thought was dodgy.

0.07 Sun Dec 2 18:41:02 200?
    - Fix test failures from CPANTS (due to us not requiring Catalyst::Devel)
    - Test a root which isn't at /, fix bugs in path resolution attaining to same ;)
    - Fix META.yml authors

0.06 Wed Nov 28 01:19:22 2007
    - Re-written to not leak connections or RAM. SVN::Client removed. Almost complete re-write 'under the hood'.
    - Test suite which tests the module for real + stress test added
    - Removed (non working) realpath method on Catalyst::Model::SVN::Item
    - Added a (horrifically inefficient) propget method to 
      Catalyst::Model::SVN::Item
    - Cleaned up path handling in cat and ls methods to conform. You can now 
      supply either a path from the repository root, or an entire URI, to both 
      methods.
    - Path and uri methods now return canonical URIs/paths always (e.g. /f1 
      not //f1 or svn://host/path rather than svn://host//path)
    - Move Catalyst::Model::SVN::Item class into it's own file.
    - Added a test for kwalitee

0.05 Sun Jun 18 16:41:25 2006
    - Added missing use DateTime statement (RT#19882)

0.04 Mon Nov 21 22:05:13 2005
    - Tweaked SYNOPSIS pod
    - Added pod notes about copy->original path resolution

0.03 Sat Nov 19 19:22:03 2005
    - Added t/catalyst_helpers_model_svn.t tests
    - Fixed various typos

0.02 Fri Nov 17 22:53:18 2005
    - Fixed not found errors when trying to read contents or logs of copies
      Big hats off to Gerard Gerritsen for pointing me in the right direction.
    - Added pod_spelling.t

0.01 Tue Nov 01 19:05:04 2005
    - Initial release


t/lib/TestLib.pm
package TestLib;
use strict;
use warnings;
use File::Temp qw( tempdir );

sub new {
    my ( $class ) = @_;
    my $self = {
        repos_uri => 'svn://localhost/',     # Eeek, default port?   
    };
    return bless $self, $class;
}

sub create_svn_repos {
    my ( $self ) = @_;
    {
        my %p;
        if (! $ENV{DEBUG_TEST}) {
            $p{CLEANUP} = 1;
        }

        $self->{tempdir} = tempdir( %p );
        warn('Created temp directory at ' . $self->{tempdir}) if $ENV{DEBUG_TEST};
    };

    chdir($self->{tempdir}) || die 'Could not change to temp directory';

    die('svnadmin failed to create repository') 
        if (system('svnadmin create repos') != 0);

    # Re-write config to make RW repos
    {
        my $CONF;
        open($CONF, '>', 'repos/conf/svnserve.conf') || die;
        print $CONF "[general]\nanon-access = read\nanon-access = write\n";
        close $CONF;
    };

    my $svnserve = `which svnserve`;
    chomp($svnserve);
    die('Cannot locate svnserve') unless $svnserve;
    warn("Created repository, svnserve is $svnserve") if $ENV{DEBUG_TEST};
    # Nasty hackery here :(
    if ($self->{svnserve_pid} = fork) {
    	# Parent, continue!
        warn("svnserve PID is " . $self->{svnserve_pid});
        sleep 3; # Nasty hack to wait for svnserve to start
    } 
    else {
    	die "cannot fork: $!" unless defined $self->{svnserve_pid};
    	$|++;
	my @cmd = ($svnserve, '-r', 'repos', '-d', '--foreground');
    	warn("Running " . join(" ", @cmd)) if $ENV{DEBUG_TEST};
	exec(@cmd) or die("Could not exec $svnserve");
    	exit;
    }
    
    my $cmd = 'svn co ' . $self->{repos_uri} . ' checkout';
    warn("Running $cmd") if $ENV{DEBUG_TEST};
    die('first checkout did not work: ' . $cmd)
        if (system($cmd));

    chdir('checkout') || die;
    mkdir('subdir') || die;

    # Revision 1
    die('Adding first subdir failed')
        if (system('svn add subdir && svn commit -m"make subdir, revision 1" subdir'));

    mkdir('subdir/s2');
    mkdir('subdir/s3');

    {
        my $F;
        open($F, '>', 'f1') || die;
        print $F "  File 1, rev 1\n  ";
        close $F;
    };

    {
        my $F;
        open($F, '>', 'subdir/f2') || die;
        print $F "File 2, rev 1\n";
        close $F;
    };

    # Revision 2
    die('Adding second subdir and 2 files failed')
        if (system('svn add subdir/s2 && svn add subdir/s3 && svn add f1 && svn add subdir/f2 && svn commit -m"make 2 more subdirs, and 2 files revision 2"'));

    # Revision 3
    die('Propset on a file failed')
        if (system('svn propset svn:mime-type "text/plain" f1 && svn commit -m"Do a propset"'));

    # Revision 4
    die('Move file failed')
        if (system('svn move subdir/f2 subdir/f2.moved && svn commit -m"Do a move"'));

    # Revision 5
    die('Move dir failed')
        if (system('svn move subdir/s3 subdir/s3.moved && svn commit -m"Do another move"'));

    warn("Finished creating repositry, returning " . $self->{repos_uri}) if $ENV{DEBUG_TEST};

    return $self->{repos_uri};
}

sub DESTROY {
    my ( $self ) = @_;
    system('kill ' . $self->{svnserve_pid});
    # FIXME - trust File::Temp to tear down temp dir
}

1;


Makefile.PL
# $Id: /mirror/claco/Catalyst-Model-SVN/trunk/Makefile.PL 4827 2008-05-26T08:51:36.007490Z bobtfish  $
use ExtUtils::MakeMaker;
use 5.008;
use strict;
use warnings;

WriteMakefile(
    NAME         => 'Catalyst::Model::SVN',
    VERSION_FROM => 'lib/Catalyst/Model/SVN.pm',
    AUTHOR       => 'Christopher H. Laco <claco@chrislaco.com>, Tomas Doran <bobtfish@bobtfish.net>',
    ABSTRACT     => 'Catalyst Model to browse Subversion repositories',
    PREREQ_PM    => {
        'Catalyst::Runtime'     => 0,
        'Catalyst::Devel'       => 0, # Build dep
        'SVN::Core'             => 0,
        'IO::Scalar'            => 0,
        'Path::Class'           => 0,
        'DateTime'              => 0,
        'IO::Scalar'            => 0,
    },
    (ExtUtils::MakeMaker->VERSION >= 6.11) ?
    (NO_META => 1) : (),
    dist => {
        PREOP => 'pod2text lib/Catalyst/Model/SVN.pm > $(DISTVNAME)/README',
    },
    clean => { FILES => 't/TestApp'},
);


lib/Catalyst/Model/SVN/Item.pm
package Catalyst::Model::SVN::Item;
use strict;
use warnings;
use Path::Class qw( file dir );
use Scalar::Util qw( blessed );
use SVN::Core;
use DateTime;
use Carp qw( confess );
use overload '""' => \&stringify, fallback => 1;

our $VERSION = '0.11';

sub new {
    my ( $class, $args ) = @_;
    confess('new needs a hashref parameter') unless ref($args) eq 'HASH';
    my $self = bless $args, ref($class) || $class;

    return $self;
}

sub author {
    return shift->{author};
}

# Name is a mandatory parameter
sub name {
    return shift->{name};
}

sub is_directory {
    my $self = shift;
    my $kind = $self->kind;

    my $is_dir = ( $kind == $SVN::Node::dir ) ? 1 : 0;
    return $is_dir;
}

sub is_file {
    my $self = shift;
    my $kind = $self->kind;

    return ( $kind == $SVN::Node::file ) ? 1 : 0;
}

# In the cat case, a kind ($SVN::Node::File) is passed in the constructor.
sub kind {
    my ($self) = @_;
    return $self->{kind};
}

sub propget {
    my ( $self, $propname ) = @_;

    confess('No propname passed') unless defined $propname;
    return $self->{props}->{$propname} if exists $self->{props};
    $self->{props}
        = ( $self->{svn}->_get_file( $self->path, $self->revision ) )
        [1];
    return $self->{props}->{$propname};
}

sub contents {
    my $self = shift;

    return $self->{contents} if exists $self->{contents};

    # FIXME - use _ra_path?
    my $location = dir( $self->{path}, $self->{name} )->stringify;
    return $self->{contents}
        = $self->{svn}->cat( $location, $self->revision );

}

sub path {
    my ( $self ) = @_;
    # FIXME - use _ra_path?
    return file( $self->{path}, $self->name )->stringify;
}

sub uri {
    my $self = shift;

    my $u = URI->new( $self->{repos} );
    $u->path( $self->path );    # Memoize?
    return $u;
}

sub log {
    my $self = shift;

    return $self->{log} if exists $self->{log};

    use Data::Dumper;

#warn("Log params " . Dumper([[$self->path], #const apr_array_header_t *paths,
#$self->revision, # svn_revnum_t start,
#$self->revision, # svn_revnum_t end,
#1, # svn_boolean_t discover_changed_paths,
#1, # svn_boolean_t strict_node_history,
#1, # svn_boolean_t include_merged_revisions,
#]));
    my ( $changes, $revision, $author, $date );
    eval {
        $self->{svn}->_ra->get_log(
            [ $self->path ],    #const apr_array_header_t *paths,
            $self->revision,    # svn_revnum_t start,
            $self->revision,    # svn_revnum_t end,
            1,                  # svn_boolean_t discover_changed_paths,
            1,                  # svn_boolean_t strict_node_history,
            1,                  # svn_boolean_t include_merged_revisions,
            sub {
                ( $changes, $revision, $author, $date, $self->{log} ) = @_;

                #warn("ITEM log Callback with params: " . Dumper(\@_));
            },    #svn_log_entry_receiver_t receiver, void *receiver_baton
        );
    };
    if ($@) {
        my $path
            = $self->{svn}->_resolve_copy( $self->path, $self->revision );

        if ( $path ne $self->path ) {
            $self->{svn}->_ra->get_log(
                [$path],            #const apr_array_header_t *paths,
                $self->revision,    # svn_revnum_t start,
                $self->revision,    # svn_revnum_t end,
                1,                  # svn_boolean_t discover_changed_paths,
                1,                  # svn_boolean_t strict_node_history,
                1,                  # svn_boolean_t include_merged_revisions,
                sub {
                    ( $changes, $revision, $author, $date, $self->{log} )
                        = @_;

                    #warn("ITEM log Callback with params: " . Dumper(\@_));
                },    #svn_log_entry_receiver_t receiver, void *receiver_baton
            );
        }
        else {
            die $@;
        }
    }
    return $self->{log};
}

sub size {
    return shift->{size};
}

sub time {
    my $self      = shift;
    my $item_time = $self->{time};

    if ( !blessed($item_time) ) {

        my $time
            = DateTime->from_epoch( epoch => substr( $item_time, 0, 10 ) );
        $time->add_duration(
            DateTime::Duration->new(
                nanoseconds => substr( $item_time, 10 )
            )
        );

        $time->set_time_zone('UTC');

        $self->{time} = $time;
    }

    return $self->{time};
}

sub revision {
    return shift->{created_rev};
}

sub stringify {
    my $self = shift;

    return $self->{name};
}

1;

__END__

=head1 NAME

Catalyst::Model::SVN::Item - Object representing a file or directory in a subversion repository.

=head1 SYNOPSIS

See L<Catalyst::Model::SVN>

=head1 DESCRIPTION

This class provides an interface to any versioned item in Subversion.

=head1 METHODS

=head2 author

The author of the latest revision of the current item.

=head2 name

Returns the name of the current item.

=head2 is_directory

Returns 1 if the current item is a directory; 0 otherwise.

=head2 is_file

Returns 1 if the current item is a file; 0 otherwise.

=head2 kind

Returns the kind  of the current item. See L<SVN::Core> for the possible types,
usually $SVN::Node::path or $SVN::Node::file.

=head2 propget ($propname) 

The property on the item named $propname

=head2 contents

The contents of the of the current item. This is the same as
calling C<Catalyst::Model::SVN->cat($item->uri, $item->revision)

=head2 path

Returns the path of the current item relative to the repository root.

=head2 uri

Returns the full repository path of the current item.

=head2 log

Returns the last log entry for the current item. Be forewarned, this makes an
extra call to the repository, which is slow. Only use this if you are listing a
single item, and not when looping through large collections of items. If the
current item is a copy, the logs are transversed to find the original. The
request is then reissued for the original path for the C<revision> specified.

=head2 size

Returns the raw file size in bytes for the current item.

=head2 time

Returns the last modified time of the current item as a L<DateTime> object.

=head2 revision

Returns the revision number of this item

=head1 SEE ALSO

L<Catalyst::Manual>, L<Catalyst::Helper>, L<Catalyst::Model::SVN>, L<SVN::Ra>

=head1 AUTHORS

    Christopher H. Laco
    CPAN ID: CLACO
    claco@chrislaco.com
    http://today.icantfocus.com/blog/

    Tomas Doran
    CPAN ID: BOBTFISH
    bobtfish@bobtfish.net



lib/Catalyst/Model/SVN.pm
# $Id: /mirror/claco/Catalyst-Model-SVN/trunk/lib/Catalyst/Model/SVN.pm 4884 2008-05-26T09:02:42.083076Z bobtfish  $
package Catalyst::Model::SVN;
use strict;
use warnings;
use SVN::Core;
use SVN::Ra;
use IO::Scalar;
use URI;
use Path::Class qw( dir file );
use NEXT;
use DateTime;
use Catalyst::Model::SVN::Item;
use Scalar::Util qw/blessed/;
use Carp qw/confess croak/;
use base 'Catalyst::Model';

our $VERSION = '0.12';

__PACKAGE__->mk_ro_accessors('repository');
__PACKAGE__->config( revision => 'HEAD' );

sub new {
    my ( $self, $c, $config ) = @_;

    $self = $self->NEXT::new(@_);

    die("No configured repository") unless $self->repository;

    my $root_pool = SVN::Pool->new_default;
    my $ra        = SVN::Ra->new(
        url  => $self->repository,
        auth => undef,
        pool => $root_pool,
    );
    
    $self->{pool} = $root_pool;
    $self->{ra}   = $ra;

    return $self;
}

sub _ra {
    my $self = shift;
    confess('Need an instance') unless blessed $self;
    return $self->{ra};
}

sub revision {
    my $self    = shift;
    my $subpool = SVN::Pool::new_default_sub;
    return $self->_ra->get_latest_revnum();
}

sub ls {
    my ( $self, $path, $revision ) = @_;
    
    $revision ||= ($self->{revision} || $self->config->{revision});
    if ( $revision eq 'HEAD' ) {
        $revision = $SVN::Core::INVALID_REVNUM;
    }
    my $subpool = SVN::Pool::new_default_sub;

    my @nodes;
    my ( $dirents, $revnum, $props )
        = $self->_ra->get_dir( _ra_path( $self, $path ), $revision );

# Note that simple data which comes back here is ok, but the dirents data structure
# will be magically deallocated when $subpool goes out of scope, so we borg all the
# info from it now..

    @nodes = map {
        Catalyst::Model::SVN::Item->new(
            {   repos       => $self->repository,
                name        => $_,
                path        => $path,
                svn         => $self,
                size        => $dirents->{$_}->size,
                kind        => $dirents->{$_}->kind,
                time        => $dirents->{$_}->time,
                author      => $dirents->{$_}->last_author,
                created_rev => $dirents->{$_}->created_rev,
            }
        );
    } sort keys %{$dirents};

    return wantarray ? @nodes : \@nodes;
}

# _ra_path( $path )
#
# Takes a path or URL, and returns a normalised from relative to the 
# configured repository path.

sub _ra_path { # FIXME - This is fugly..
    my ( $self, $path ) = @_;
    $path ||= '/';
    my $uri = URI->new($path);
    $path =~ s|//+|/|;
    my $ra_path;
    if ($uri->scheme) {
        # Was full URI
        $ra_path = file( $uri->path );
    }
    else {
        $ra_path = file( URI->new($self->repository)->path, $path );
    }
    
    $ra_path = $ra_path->stringify;
    $ra_path =~ s|/$||; # Remove trailing / or svn can crash
    
    return $ra_path;
}

sub cat {
    my ( $self, $path, $revision ) = @_;
    return ( $self->_get_file( $path, $revision ) )[0];
}

sub propget {
    my ( $self, $path, $propname, $revision ) = @_;

    croak('No propname passed to propget method') unless defined $propname;
    
    my $props_hr = $self->properties_hr($path, $revision);
    return $props_hr->{$propname}
}

sub properties_hr {
    my ( $self, $path, $revision ) = @_;

    croak('No path passed to props_hr method') unless defined $path;
    
    return ( $self->_get_file( $path, $revision ) )[1];
}

# _get_file( $path [, $revision] )
#
# Calls the L<SVN::Ra> get_file method. Handles directories and files which 
# have moved in older revisions

sub _get_file {
    my ( $self, $path, $revision ) = @_;
    my $repos_path = _ra_path( $self, $path );
    $revision = undef if ( defined $revision && $revision eq 'HEAD' );
    $revision ||= $SVN::Core::INVALID_REVNUM;
    my $requested_path = $repos_path;
    my $file           = IO::Scalar->new;
    my $subpool        = SVN::Pool::new_default_sub;
    my ( $revnum, $props );
    use Data::Dumper;
    eval {
        ( $revnum, $props )
            = $self->_ra->get_file( $repos_path, $revision, $file );

    };
    return ( $file, $props ) unless $@;

    # Handle dictionary case..
    if ( $@ =~ /Attempted to get checksum of a \*non\*-file node/ ) {
        return;
    }

    if ( $@ =~ /ile not found/ ) {
        $repos_path = $self->_resolve_copy( $repos_path, $revision );

        if ( $repos_path ne $requested_path ) {
            return $self->_get_file( $repos_path, $revision );
        }
    }

    die $@;
}

sub _resolve_copy {
    my ( $self, $path, $revision ) = @_;
    my $subpool = SVN::Pool::new_default_sub;

    my $copyfrom;
    $self->_ra->get_log(
        [$path],                       # const apr_array_header_t *paths,
        $self->_ra->get_latest_revnum, # svn_revnum_t start,
        $revision,                     # svn_revnum_t end,
        1,                             # svn_boolean_t discover_changed_paths,
        1,                             # svn_boolean_t strict_node_history,
        1,       # svn_boolean_t include_merged_revisions,
        sub {    # svn_log_entry_receiver_t receiver, void *receiver_baton
            return if $copyfrom;
            my $changes = shift;
            foreach my $change ( keys %$changes ) {
                my $obj    = $changes->{$change};
                my $action = $obj->action;
                $copyfrom = $obj->copyfrom_path;
                if ( $obj->action eq 'A' && $copyfrom ) {
                    $path =~ s/$change/$copyfrom/;
                    last;
                }
            }
        },
    );
    return $path;
}

1;

__END__

=head1 NAME

Catalyst::Model::SVN - Catalyst Model to browse Subversion repositories

=head1 SYNOPSIS

    # Model
    __PACKAGE__->config(
        repository => '/path/to/svn/root/or/path'
    );

    # Controller
    sub default : Private {
        my ($self, $c) = @_;
        my $path = join('/', $c->req->args);
        my $revision = $c->req->param('revision') || 'HEAD';

        $c->stash->{'repository_revision'} = MyApp::M::SVN->revision;
        $c->stash->{'items'} = MyApp::M::SVN->ls($path, $revision);

        $c->stash->{'template'} = 'blog.tt';
    };

=head1 DESCRIPTION

This model class uses the perl-subversion bindings to access a Subversion
repository and list items and view their contents. It is currently only a
read-only client but may expand to be a fill fledged client at a later time.

=head1 CONFIG

The following configuration options are available:

=head2 repository

Returns a URI object of the full path to the root of, or any directory in your Subversion
repository. This can be one of http://, svn://, or file:/// schemes. 

This value comes from the config key 'repository'.

=head2 revision

This is the default revision to use when no revision is specified. By default,
this will be C<HEAD>.

=head1 METHODS

=head2 cat($path [, $revision])

Returns the contents of the path specified. If C<path> is a copy, the logs are
transversed to find original. The request is then reissued for the original path
for the C<revision> specified.

=head2 ls($path [, $revision])

Returns a array of L<Catalyst::Model::SVN::Item> objects in list context, each
representing an entry in the specified repository path. In scalar context, it
returns an array reference.  If C<path> is a copy, the logs are
transversed to find the original. The request is then reissued for the original
path for the C<revision> specified.

=head2 propget($path, $propname [, $revision])

Returns a specific property for a path at a specified revision name.

Note: This method is inefficient, if you want to extract multiple properties 
of a single item then use the props_hr method.

=head2 properties_hr($path [, $revision])

Returns a reference to a hash with all the properties set on an object at a specific revision.

=head2 repository

Returns the repository specified in the configuration C<repository> option.

=head2 revision

Returns the latest revisions of the repository you are connected to.

=head1 SEE ALSO

L<Catalyst::Manual>, L<Catalyst::Helper>, L<Catalyst::Model::SVN::Item>, L<SVN::Ra>

=head1 AUTHORS

    Christopher H. Laco
    CPAN ID: CLACO
    claco@chrislaco.com
    http://today.icantfocus.com/blog/
    
    Tomas Doran
    CPAN ID: BOBTFISH
    bobtfish@bobtfish.net
    
